{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4cc4c9a5-b7de-4298-91ed-57dfe07c2aea",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "import pandas as pd\n",
    "import json\n",
    "import re, os\n",
    "import pyBigWig\n",
    "import glob\n",
    "\n",
    "HEADER_HASH = '#'\n",
    "hgnc = pyBigWig.open(\"project/aux/hgnc.bb\")\n",
    "    \n",
    "def is_in_gene_set(pos,gene_list):\n",
    "    is_gene = False\n",
    "    for gene in gene_list:\n",
    "        chromo = gene.split(':')[0]\n",
    "        if pos['chromosome'] != chromo: continue\n",
    "        index = gene.split(':')[1]\n",
    "        start = index.split('-')[0]\n",
    "        end = index.split('-')[1]\n",
    "        if pos['position'] >= int(start) and pos['position'] <= int(end):\n",
    "            is_gene = True\n",
    "            break\n",
    "    return is_gene\n",
    "\n",
    "def get_gene_name(chrom,pos):\n",
    "    return hgnc.entries(chrom,pos,pos+1)[0][2].split('\\t')[6], hgnc.entries(chrom,pos,pos+1)[0][2].split('\\t')[7]\n",
    "    \n",
    "def parse_annotated_json(path, gene_subset):\n",
    "    #Takes in a annotated VCF file and filters only variants that lie within a gene subset\n",
    "    #Gene_subset is expected to be a list of genome positions in the format of chr:start-end\n",
    "    header = ''\n",
    "    positions = []\n",
    "    genes = []\n",
    "    is_header_line = True\n",
    "    is_position_line = False\n",
    "    is_gene_line = False\n",
    "    gene_section_line = '],\"genes\":['\n",
    "    end_line = ']}'\n",
    "    with gzip.open(file, 'rt') as f:\n",
    "        position_count = 0\n",
    "        gene_count = 0\n",
    "        count = 0\n",
    "        for line in f:\n",
    "            trim_line = line.strip()\n",
    "            if is_header_line:\n",
    "                ## only keep the \"header\" field content from the line\n",
    "                header = trim_line[10:-14]\n",
    "                is_header_line = False\n",
    "                is_position_line = True\n",
    "            elif trim_line == gene_section_line:\n",
    "                is_gene_line = True\n",
    "                is_position_line = False\n",
    "                continue\n",
    "            elif trim_line == end_line:\n",
    "                break\n",
    "            else:\n",
    "                if is_position_line:\n",
    "                    ## remove the trailing ',' if there is\n",
    "                    trim_line = trim_line.rstrip(',')\n",
    "                    pos = json.loads(trim_line)\n",
    "                    if is_in_gene_set(pos,gene_subset):\n",
    "                        positions.append(trim_line)\n",
    "                        position_count += 1\n",
    "                    #if position_count == 100000: break\n",
    "                if is_gene_line:\n",
    "                    ## remove the trailing ',' if there is\n",
    "                    genes.append(trim_line.rstrip(','))\n",
    "                    gene_count += 1\n",
    "    return {'header':header, 'positions':positions, 'genes':genes}\n",
    "\n",
    "def get_clingen_annot(records):\n",
    "    labels = []\n",
    "    for rec in records:\n",
    "        labels.append(rec['clinicalInterpretation'])\n",
    "    return(max(set(labels), key=labels.count))\n",
    "def get_clinvar_annot(records):\n",
    "    labels = []\n",
    "    for rec in records:\n",
    "        labels = labels + rec['significance']\n",
    "    return(max(set(labels), key=labels.count))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "dbe5653c-820d-489d-b86f-3a545be02b81",
   "metadata": {},
   "outputs": [],
   "source": [
    "files = glob.glob('project/results/germline/*/*/*.sv.vcf.annotated.json.gz')\n",
    "gene_set = pd.read_csv(\"project/aux/Mito-Lyso-Pesticide_PD_genes.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b71157c9-791d-40d2-a8ba-38803e5f6fb4",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[3], line 4\u001b[0m\n\u001b[1;32m      2\u001b[0m df \u001b[38;5;241m=\u001b[39m []\n\u001b[1;32m      3\u001b[0m sample_name \u001b[38;5;241m=\u001b[39m file\u001b[38;5;241m.\u001b[39msplit(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m/\u001b[39m\u001b[38;5;124m'\u001b[39m)[\u001b[38;5;241m4\u001b[39m]\n\u001b[0;32m----> 4\u001b[0m annotated_vcf \u001b[38;5;241m=\u001b[39m \u001b[43mparse_annotated_json\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfile\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mgene_set\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpos\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m      5\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m pos \u001b[38;5;129;01min\u001b[39;00m annotated_vcf[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mpositions\u001b[39m\u001b[38;5;124m'\u001b[39m]:\n\u001b[1;32m      6\u001b[0m     x \u001b[38;5;241m=\u001b[39m json\u001b[38;5;241m.\u001b[39mloads(pos)\n",
      "Cell \u001b[0;32mIn[1], line 42\u001b[0m, in \u001b[0;36mparse_annotated_json\u001b[0;34m(path, gene_subset)\u001b[0m\n\u001b[1;32m     40\u001b[0m gene_count \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m0\u001b[39m\n\u001b[1;32m     41\u001b[0m count \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m0\u001b[39m\n\u001b[0;32m---> 42\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m line \u001b[38;5;129;01min\u001b[39;00m f:\n\u001b[1;32m     43\u001b[0m     trim_line \u001b[38;5;241m=\u001b[39m line\u001b[38;5;241m.\u001b[39mstrip()\n\u001b[1;32m     44\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m is_header_line:\n\u001b[1;32m     45\u001b[0m         \u001b[38;5;66;03m## only keep the \"header\" field content from the line\u001b[39;00m\n",
      "File \u001b[0;32m/usr/lib/python3.10/gzip.py:314\u001b[0m, in \u001b[0;36mGzipFile.read1\u001b[0;34m(self, size)\u001b[0m\n\u001b[1;32m    312\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m size \u001b[38;5;241m<\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[1;32m    313\u001b[0m     size \u001b[38;5;241m=\u001b[39m io\u001b[38;5;241m.\u001b[39mDEFAULT_BUFFER_SIZE\n\u001b[0;32m--> 314\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_buffer\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread1\u001b[49m\u001b[43m(\u001b[49m\u001b[43msize\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/usr/lib/python3.10/_compression.py:68\u001b[0m, in \u001b[0;36mDecompressReader.readinto\u001b[0;34m(self, b)\u001b[0m\n\u001b[1;32m     66\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mreadinto\u001b[39m(\u001b[38;5;28mself\u001b[39m, b):\n\u001b[1;32m     67\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mmemoryview\u001b[39m(b) \u001b[38;5;28;01mas\u001b[39;00m view, view\u001b[38;5;241m.\u001b[39mcast(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mB\u001b[39m\u001b[38;5;124m\"\u001b[39m) \u001b[38;5;28;01mas\u001b[39;00m byte_view:\n\u001b[0;32m---> 68\u001b[0m         data \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mlen\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mbyte_view\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     69\u001b[0m         byte_view[:\u001b[38;5;28mlen\u001b[39m(data)] \u001b[38;5;241m=\u001b[39m data\n\u001b[1;32m     70\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(data)\n",
      "File \u001b[0;32m/usr/lib/python3.10/gzip.py:494\u001b[0m, in \u001b[0;36m_GzipReader.read\u001b[0;34m(self, size)\u001b[0m\n\u001b[1;32m    491\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_new_member \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mFalse\u001b[39;00m\n\u001b[1;32m    493\u001b[0m \u001b[38;5;66;03m# Read a chunk of data from the file\u001b[39;00m\n\u001b[0;32m--> 494\u001b[0m buf \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_fp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[43mio\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mDEFAULT_BUFFER_SIZE\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    496\u001b[0m uncompress \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_decompressor\u001b[38;5;241m.\u001b[39mdecompress(buf, size)\n\u001b[1;32m    497\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_decompressor\u001b[38;5;241m.\u001b[39munconsumed_tail \u001b[38;5;241m!=\u001b[39m \u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n",
      "File \u001b[0;32m/usr/lib/python3.10/gzip.py:97\u001b[0m, in \u001b[0;36m_PaddedFile.read\u001b[0;34m(self, size)\u001b[0m\n\u001b[1;32m     94\u001b[0m read \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_read\n\u001b[1;32m     95\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_read \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m     96\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_buffer[read:] \u001b[38;5;241m+\u001b[39m \\\n\u001b[0;32m---> 97\u001b[0m        \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfile\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[43msize\u001b[49m\u001b[38;5;241;43m-\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_length\u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "for file in files:\n",
    "    df = []\n",
    "    sample_name = file.split('/')[4]\n",
    "    annotated_vcf = parse_annotated_json(file, gene_set.pos)\n",
    "    for pos in annotated_vcf['positions']:\n",
    "        x = json.loads(pos)\n",
    "        for var_dict in x['variants']:\n",
    "            build_dict = {'vid' : var_dict['vid'],\n",
    "                          'chromosome' : var_dict['chromosome'],\n",
    "                          'begin' : var_dict['begin'],\n",
    "                          'end' : var_dict['end'],\n",
    "                          'gene_symbol' : get_gene_name(var_dict['chromosome'],int(var_dict['begin']))[0],\n",
    "                          'gene_name' : get_gene_name(var_dict['chromosome'],int(var_dict['begin']))[1],\n",
    "                          'refAllele' : var_dict['refAllele'],\n",
    "                          'altAllele' : var_dict['altAllele'],\n",
    "                          'variantType' : var_dict['variantType'],\n",
    "                          'quality' : x['quality'],\n",
    "                          'samples' : x['samples']\n",
    "                         }\n",
    "            if 'svLength' in x:\n",
    "                build_dict = build_dict | {'svLength' : x['svLength']}\n",
    "            if 'clingen' in x:\n",
    "                build_dict = build_dict | {'clingen' : get_clingen_annot(x['clingen'])}\n",
    "            if 'clinvar' in x:\n",
    "                build_dict = build_dict | {'clinvar' : get_clinvar_annot(x['clinvar'])}\n",
    "            if 'clinvar' in var_dict: # Alot of information...unsure of how to simplify into tabular. Include all for now\n",
    "                build_dict = build_dict | {'clinvar' : var_dict['clinvar']}\n",
    "            if 'dbsnp' in var_dict:\n",
    "                build_dict = build_dict | {'dbsnp' : var_dict['dbsnp']}\n",
    "            if 'hgvsg' in var_dict:\n",
    "                build_dict = build_dict | {'hgvsg' : var_dict['hgvsg']}\n",
    "            if 'gnomad' in var_dict:\n",
    "                prefixed_dict = {'gnomad_'+k: v for k, v in var_dict['gnomad'].items()}\n",
    "                build_dict = build_dict | prefixed_dict\n",
    "            if 'regulatoryRegions' in var_dict:\n",
    "                prefixed_dict = {'regulatoryRegions_'+k: v for k, v in var_dict['regulatoryRegions'][0].items()}\n",
    "                build_dict = build_dict | prefixed_dict\n",
    "            if 'cosmic' in var_dict:\n",
    "                build_dict = build_dict | {'cosmic' : var_dict['cosmic']}\n",
    "            if 'isStructuralVariant' in var_dict:\n",
    "                build_dict = build_dict | {'isStructuralVariant' : var_dict['isStructuralVariant']}\n",
    "            df.append(build_dict)\n",
    "    df = pd.DataFrame(df)\n",
    "    df.to_csv(f'germline_variants/sv/{sample_name}_germline_sv.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3011628f-c9b3-4634-8554-4aa3bbfea116",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
