{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4cc4c9a5-b7de-4298-91ed-57dfe07c2aea",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "import pandas as pd\n",
    "import json\n",
    "import re, os\n",
    "import pyBigWig\n",
    "import glob\n",
    "\n",
    "HEADER_HASH = '#'\n",
    "hgnc = pyBigWig.open(\"project/aux/hgnc.bb\")\n",
    "    \n",
    "def is_in_gene_set(pos,gene_list):\n",
    "    is_gene = False\n",
    "    for gene in gene_list:\n",
    "        chromo = gene.split(':')[0]\n",
    "        if pos['chromosome'] != chromo: continue\n",
    "        index = gene.split(':')[1]\n",
    "        start = index.split('-')[0]\n",
    "        end = index.split('-')[1]\n",
    "        if pos['position'] >= int(start) and pos['position'] <= int(end):\n",
    "            is_gene = True\n",
    "            break\n",
    "    return is_gene\n",
    "\n",
    "def get_gene_name(chrom,pos):\n",
    "    return hgnc.entries(chrom,pos,pos+1)[0][2].split('\\t')[6], hgnc.entries(chrom,pos,pos+1)[0][2].split('\\t')[7]\n",
    "    \n",
    "def parse_annotated_json(path, gene_subset):\n",
    "    #Takes in a annotated VCF file and filters only variants that lie within a gene subset\n",
    "    #Gene_subset is expected to be a list of genome positions in the format of chr:start-end\n",
    "    header = ''\n",
    "    positions = []\n",
    "    genes = []\n",
    "    is_header_line = True\n",
    "    is_position_line = False\n",
    "    is_gene_line = False\n",
    "    gene_section_line = '],\"genes\":['\n",
    "    end_line = ']}'\n",
    "    with gzip.open(file, 'rt') as f:\n",
    "        position_count = 0\n",
    "        gene_count = 0\n",
    "        count = 0\n",
    "        for line in f:\n",
    "            trim_line = line.strip()\n",
    "            if is_header_line:\n",
    "                ## only keep the \"header\" field content from the line\n",
    "                header = trim_line[10:-14]\n",
    "                is_header_line = False\n",
    "                is_position_line = True\n",
    "            elif trim_line == gene_section_line:\n",
    "                is_gene_line = True\n",
    "                is_position_line = False\n",
    "                continue\n",
    "            elif trim_line == end_line:\n",
    "                break\n",
    "            else:\n",
    "                if is_position_line:\n",
    "                    ## remove the trailing ',' if there is\n",
    "                    trim_line = trim_line.rstrip(',')\n",
    "                    pos = json.loads(trim_line)\n",
    "                    if is_in_gene_set(pos,gene_subset):\n",
    "                        positions.append(trim_line)\n",
    "                        position_count += 1\n",
    "                    #if position_count == 100000: break\n",
    "                if is_gene_line:\n",
    "                    ## remove the trailing ',' if there is\n",
    "                    genes.append(trim_line.rstrip(','))\n",
    "                    gene_count += 1\n",
    "    return {'header':header, 'positions':positions, 'genes':genes}\n",
    "\n",
    "def get_clingen_annot(records):\n",
    "    labels = []\n",
    "    for rec in records:\n",
    "        labels.append(rec['clinicalInterpretation'])\n",
    "    return(max(set(labels), key=labels.count))\n",
    "def get_clinvar_annot(records):\n",
    "    labels = []\n",
    "    for rec in records:\n",
    "        labels = labels + rec['significance']\n",
    "    return(max(set(labels), key=labels.count))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "dbe5653c-820d-489d-b86f-3a545be02b81",
   "metadata": {},
   "outputs": [],
   "source": [
    "files = glob.glob('project/results/germline/*/*/*.cnv.vcf.annotated.json.gz')\n",
    "gene_set = pd.read_csv(\"project/aux/Mito-Lyso-Pesticide_PD_genes.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b71157c9-791d-40d2-a8ba-38803e5f6fb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "for file in files:\n",
    "    df = []\n",
    "    sample_name = file.split('/')[4]\n",
    "    annotated_vcf = parse_annotated_json(file, gene_set.pos)\n",
    "    for pos in annotated_vcf['positions']:\n",
    "        x = json.loads(pos)\n",
    "        for var_dict in x['variants']:\n",
    "            build_dict = {'vid' : var_dict['vid'],\n",
    "                          'chromosome' : var_dict['chromosome'],\n",
    "                          'begin' : var_dict['begin'],\n",
    "                          'end' : var_dict['end'],\n",
    "                          'gene_symbol' : get_gene_name(var_dict['chromosome'],int(var_dict['begin']))[0],\n",
    "                          'gene_name' : get_gene_name(var_dict['chromosome'],int(var_dict['begin']))[1],\n",
    "                          'refAllele' : var_dict['refAllele'],\n",
    "                          'altAllele' : var_dict['altAllele'],\n",
    "                          'variantType' : var_dict['variantType'],\n",
    "                          'samples' : x['samples']}\n",
    "            if 'quality' in x:\n",
    "                build_dict = build_dict | {'quality' : x['quality']}\n",
    "            if 'svEnd' in x:\n",
    "                build_dict = build_dict | {'svEnd' : x['svEnd']}\n",
    "            if 'clingen' in x:\n",
    "                build_dict = build_dict | {'clingen' : get_clingen_annot(x['clingen'])}\n",
    "            if 'clinvar' in x:\n",
    "                build_dict = build_dict | {'clinvar' : get_clinvar_annot(x['clinvar'])}\n",
    "            if 'gnomAD-preview' in x:\n",
    "                build_dict = build_dict | {'gnomAD-preview' : x['gnomAD-preview']}\n",
    "            if 'oneKg' in var_dict:\n",
    "                prefixed_dict = {'oneKg_'+k: v for k, v in var_dict['oneKg'].items()}\n",
    "                build_dict = build_dict | prefixed_dict\n",
    "            if 'isStructuralVariant' in var_dict:\n",
    "                build_dict = build_dict | {'isStructuralVariant' : var_dict['isStructuralVariant']}\n",
    "            df.append(build_dict)\n",
    "    df = pd.DataFrame(df)\n",
    "    df.to_csv(f'germline_variants/cnv/{sample_name}_germline_cnv.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3011628f-c9b3-4634-8554-4aa3bbfea116",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
