{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5376050b-eb46-4278-b344-620dbafb98cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "import re\n",
    "import pandas as pd\n",
    "import gzip\n",
    "from cyvcf2 import VCF\n",
    "\n",
    "def get_dosage_int(geno):\n",
    "    x = re.split('\\\\||/',geno)\n",
    "    return sum([int(i) for i in x])\n",
    "\n",
    "nucleotide_map = {\n",
    "    'A':'T',\n",
    "    'T':'A',\n",
    "    'C':'G',\n",
    "    'G':'C'\n",
    "}\n",
    "def flip_dosage(num):\n",
    "    if num == 0:\n",
    "        return 2\n",
    "    elif num == 2:\n",
    "        return 0\n",
    "    else:\n",
    "        return num\n",
    "\n",
    "def fix_effect_allele(var, effect_allele, dosage_int):\n",
    "    if effect_allele == var.ALT[0]: #Matched, dont change anything\n",
    "        return dosage_int\n",
    "    if effect_allele == var.REF: #Only flip if not palindromic with reference\n",
    "        if nucleotide_map[effect_allele] == var.ALT[0]:\n",
    "            #Is palindromic, cannot use this SNP to calculate PRS\n",
    "            return 0\n",
    "        else:\n",
    "            return flip_dosage(dosage_int)\n",
    "\n",
    "def calc_prs(vcf_path,PRS_table):\n",
    "    PRS_total = []\n",
    "    vcf = VCF(vcf_path)\n",
    "    for i,row in PRS_table.iterrows():\n",
    "        iterator = vcf(f'{row[\"hg38_chr\"]}:{row[\"hg38_pos\"]}-{row[\"hg38_pos\"]}')\n",
    "        if all(False for _ in iterator):\n",
    "            if row['effect_is_ref']:\n",
    "                PRS_total.append(row['effect_weight']*2)\n",
    "            continue\n",
    "        iterator = vcf(f'{row[\"hg38_chr\"]}:{row[\"hg38_pos\"]}-{row[\"hg38_pos\"]}')\n",
    "        var = next(iterator)\n",
    "        if var.POS != row[\"hg38_pos\"]: #Sometimes vcf gets neighbour positions\n",
    "            continue\n",
    "        hg38_pos_id = f'{var.CHROM}_{var.POS}'\n",
    "        dosage_int = sum(var.genotypes[0][0:2])\n",
    "        effect_allele = PRS_table[PRS_table.hg38_pos_id==hg38_pos_id]['effect_allele'].item()\n",
    "        effect_weight = PRS_table[PRS_table.hg38_pos_id==hg38_pos_id]['effect_weight'].item()\n",
    "        fixed_dosage = fix_effect_allele(var,effect_allele,dosage_int)\n",
    "        PRS = effect_weight*fixed_dosage\n",
    "        PRS_total.append(PRS)\n",
    "    return(sum(PRS_total))\n",
    "\n",
    "def calc_mlc(vcf_path,MLC_table):\n",
    "    vcf = VCF(vcf_path)\n",
    "    MLC_scores = []\n",
    "    for v in vcf('chrM'):\n",
    "        #Skip indels\n",
    "        if len(v.ALT[0])>1 or len(v.REF)>1:\n",
    "            continue\n",
    "        #Check ref\n",
    "        if v.REF != MLC_table.loc[MLC_table.Position==v.POS,'Reference'].to_list()[0]:\n",
    "            print(v.POS)\n",
    "            print(\"Incorrect reference allele\")\n",
    "            continue\n",
    "        score = MLC_table[(MLC_table.Position==v.POS)&(MLC_table.Alternate==v.ALT[0])]['MLC_score'].item()\n",
    "        MLC_scores.append(score)\n",
    "    return(MLC_scores)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "24d4e372-f21b-43ff-a5a7-9315a346bcbd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Ploidy\n",
    "ploidy_files = glob.glob(\"project/results/germline/*/*/*.ploidy_estimation_metrics.csv\")\n",
    "dat_list = []\n",
    "for file in ploidy_files:\n",
    "    sample_name = file.split('/')[4]\n",
    "    dat = pd.read_csv(file, names = ['1','2','index',sample_name], usecols = ['index',sample_name])\n",
    "    dat = dat.set_index('index')\n",
    "    dat_list.append(dat)\n",
    "ploidy_merged_dat = pd.concat(dat_list, axis=1).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9c09bd0d-8be4-41f0-97dd-bb3a08f059c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mitochondrial Copy Number \n",
    "contig_cov_files = glob.glob(\"project/results/germline/*/*/*.wgs_contig_mean_cov.csv\")\n",
    "copy_number_dat = []\n",
    "for file in contig_cov_files:\n",
    "    sample_name = file.split('/')[4]\n",
    "    dat = pd.read_csv(file, names = ['contig','read_count','coverage'])\n",
    "    autosomal_coverage = dat[dat.contig.str.contains(\"Autosomal regions\")]['coverage'].item()\n",
    "    mitochondrial_coverage = dat[dat.contig==\"chrM\"]['coverage'].item()\n",
    "    mito_copy_number = 2*mitochondrial_coverage/autosomal_coverage\n",
    "    copy_number_dat.append([sample_name,autosomal_coverage,mitochondrial_coverage,mito_copy_number])\n",
    "copy_number_dat = pd.DataFrame(copy_number_dat, columns = ['index','autosomal_cov','mitochondrial_cov','mitochondrial_copy_number'])\n",
    "copy_number_dat = copy_number_dat.set_index('index')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "056fa72c-4b5a-447f-b64d-155ac3379cae",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Polygenic Risk Score (PRS)\n",
    "PRS_table = pd.read_csv('project/aux/PGS000902_hg38.txt', sep='\\t', comment = '#')\n",
    "vcf_files = glob.glob(\"project/results/germline/*/*/*.hard-filtered.vcf.gz\")\n",
    "prs_dat_list = []\n",
    "for vcf in vcf_files:\n",
    "    sample_name = vcf.split('/')[4]\n",
    "    prs_dat_list.append([sample_name,calc_prs(vcf,PRS_table)])\n",
    "prs_dat = pd.DataFrame(prs_dat_list, columns = ['index','prs'])\n",
    "prs_dat = prs_dat.set_index('index')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05587a57-c850-429c-9358-e9020af5eefb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mitochondrial local constraint (MLC)\n",
    "vcf_files = glob.glob(\"project/results/germline/*/*/*.hard-filtered.vcf.gz\")\n",
    "MLC_table = pd.read_csv(\"project/aux/MLC_supplementary_dataset_7.tsv\", sep='\\t')\n",
    "mlc_dat_list = []\n",
    "for vcf in vcf_files:\n",
    "    sample_name = vcf.split('/')[4]\n",
    "    MLC_scores = calc_mlc(vcf,MLC_table)\n",
    "    MSS = sum(MLC_scores)\n",
    "    mean_MLC = sum(MLC_scores)/len(MLC_scores)\n",
    "    mlc_dat_list.append([sample_name,MSS,mean_MLC])\n",
    "mlc_dat = pd.DataFrame(mlc_dat_list, columns = ['index','MSS','Mean_MLC'])\n",
    "mlc_dat = mlc_dat.set_index('index')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c34797e0-67aa-48b0-8c28-80ff03789d25",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Tumour Mutational Burden\n",
    "tmb_files = glob.glob(\"project/results/somatic/*TMB30*/*/*.tmb.metrics.csv\")\n",
    "tmb_dat_list = []\n",
    "for file in tmb_files:\n",
    "    sample_name = file.split('/')[4]\n",
    "    dat = pd.read_csv(file, names = ['1','2','index',sample_name], usecols = ['index',sample_name])\n",
    "    dat = dat.set_index('index')\n",
    "    tmb_dat_list.append(dat)\n",
    "tmb_merged_dat = pd.concat(tmb_dat_list, axis=1).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e7ceda98-ed60-41a0-be3a-08df54ab6492",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_merged = ploidy_merged_dat.merge(copy_number_dat, left_index=True, right_index=True)\n",
    "all_merged = all_merged.merge(prs_dat, left_index=True, right_index=True)\n",
    "all_merged = all_merged.merge(mlc_dat, left_index=True, right_index=True)\n",
    "all_merged = all_merged.merge(tmb_merged_dat, how = 'left', left_index=True, right_index=True)\n",
    "all_merged.to_csv('all_sample_metrics.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74622aeb-35b8-4e9d-8110-9cc420ffcd1c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
